include/ginkgo/core/base/mpi.hpp

```diff
@@ -0,0 +1,1251 @@
+    }
+
+    communicator(communicator& other)
+    {
+        GKO_ASSERT_NO_MPI_ERRORS(MPI_Comm_dup(other.comm_, &this->comm_));
+        this->size_ = get_num_ranks();
+        this->rank_ = get_my_rank();
+        this->local_rank_ = get_local_rank();
+    }
+
+    communicator& operator=(const communicator& other)
+    {
+        GKO_ASSERT_NO_MPI_ERRORS(MPI_Comm_dup(other.comm_, &this->comm_));
+        this->size_ = get_num_ranks();
+        this->rank_ = get_my_rank();
+        this->local_rank_ = get_local_rank();
+        return *this;
+    }
+
+    communicator(communicator&& other)
+    {
+        GKO_ASSERT_NO_MPI_ERRORS(MPI_Comm_dup(other.comm_, &this->comm_));
+        this->size_ = get_num_ranks();
+        this->rank_ = get_my_rank();
+        this->local_rank_ = get_local_rank();
+        other.comm_ = MPI_COMM_NULL;
+        other.size_ = 0;
+        other.rank_ = -1;
+    }
+
+    communicator& operator=(communicator&& other)
+    {
+        GKO_ASSERT_NO_MPI_ERRORS(MPI_Comm_dup(other.comm_, &this->comm_));
+        this->size_ = get_num_ranks();
+        this->rank_ = get_my_rank();
+        this->local_rank_ = get_local_rank();
+        other.size_ = 0;
+        other.rank_ = -1;
+        return *this;
+    }
+
+    static MPI_Comm get_comm_world() { return MPI_COMM_WORLD; }
+
+    static std::shared_ptr<communicator> create_world()
+    {
+        return std::make_shared<communicator>(get_comm_world());
+    }
+
+    MPI_Comm get() const { return comm_; }
+
+    int size() const { return size_; }
+
+    int rank() const { return rank_; };
+
+    int local_rank() const { return local_rank_; };
+
+    bool compare(const MPI_Comm& other) const
+    {
+        int flag;
+        GKO_ASSERT_NO_MPI_ERRORS(MPI_Comm_compare(this->comm_, other, &flag));
+        return flag;
+    }
+
+    bool operator==(const communicator& rhs) { return compare(rhs.get()); }
+
+    ~communicator()
+    {
+        if (this->comm_ && this->comm_ != MPI_COMM_NULL) {
+            MPI_Comm_free(&this->comm_);
+        }
+    }
+
+private:
+    MPI_Comm comm_;
+    int size_{};
+    int rank_{};
+    int local_rank_{};
+
+    int get_my_rank()
+    {
+        int my_rank = 0;
+        GKO_ASSERT_NO_MPI_ERRORS(MPI_Comm_rank(comm_, &my_rank));
+        return my_rank;
+    }
+
+    int get_local_rank()
+    {
+        MPI_Comm local_comm;
+        int rank;
+        GKO_ASSERT_NO_MPI_ERRORS(MPI_Comm_split_type(
+            comm_, MPI_COMM_TYPE_SHARED, 0, MPI_INFO_NULL, &local_comm));
+        GKO_ASSERT_NO_MPI_ERRORS(MPI_Comm_rank(local_comm, &rank));
+        MPI_Comm_free(&local_comm);
+        return rank;
+    }
+
+    int get_num_ranks()
+    {
+        int size = 1;
+        GKO_ASSERT_NO_MPI_ERRORS(MPI_Comm_size(comm_, &size));
+        return size;
+    }
+};
+
+
+/**
+ * Get the rank in the communicator of the calling process.
+ *
+ * @param comm  the communicator
+ */
+static double get_walltime() { return MPI_Wtime(); }
```

Comment:
also for all other non-member functions:
```suggestion
inline double get_walltime() { return MPI_Wtime(); }
```

Reply:
`inline` does not necessarily inline the functions, so I think this might give some multiple definition errors.

Reply:
You might be mixing up things here - `static` means internal linkage, so it will create the same function in each compilation unit, but not emit symbols for them. `inline` means weak linkage, so it will create the same function in each compilation unit with a weak symbol attached. As long as the function is one of the two, it will not cause multiple definition errors. `static` rarely makes sense for non-member functions in header files.

Reply:
I see. Thanks for the clarification.